pkg/controller/replication/replication_controller.go

```diff
@@ -580,17 +639,48 @@ func (rm *ReplicationManager) syncReplicationController(key string) error {
 	trace.Step("ReplicationController restored")
 	rcNeedsSync := rm.expectations.SatisfiedExpectations(rcKey)
 	trace.Step("Expectations restored")
-	podList, err := rm.podStore.Pods(rc.Namespace).List(labels.Set(rc.Spec.Selector).AsSelector())
 	if err != nil {
 		glog.Errorf("Error getting pods for rc %q: %v", key, err)
 		rm.queue.Add(key)
 		return err
 	}
 	trace.Step("Pods listed")
 
-	// TODO: Do this in a single pass, or use an index.
-	filteredPods := controller.FilterActivePods(podList.Items)
-	if rcNeedsSync {
+	// TODO: Do the List and Filter in a single pass, or use an index.
+	var filteredPods []*api.Pod
+	if rm.garbageCollectorEnabled {
+		// list all pods to include the pods that don't match the rc's selector
+		// anymore but has the stale controller ref.
+		podList, err := rm.podStore.Pods(rc.Namespace).List(labels.Everything())
+		if err != nil {
+			glog.Errorf("Error getting pods for rc %q: %v", key, err)
+			rm.queue.Add(key)
+			return err
+		}
+		var abandonPods []*api.Pod
+		filteredPods, abandonPods = controller.ControlActivePods(
+			rm.podControl, rc.ObjectMeta, labels.Set(rc.Spec.Selector).AsSelector(),
+			v1.SchemeGroupVersion.WithKind("ReplicationController"), podList.Items)
+
+		// remove the controllerRef for the pods that no longer have matching labels
+		errors := rm.removeControllerRef(abandonPods, rc.UID)
```

Comment:
It's weird that some controller ref functions are plain functions, and this one is on the replication manager. Suggest grouping them into one class. If they're always going to be done one after the other from all other controllers (whenever they support owner ref) suggest one function.